File: src/01-easy/happy-numbers/solution.rb
STDOUT.sync = true
n = gets.to_i
n.times do
    s = gets.strip
    x = s.chars.sum { |c| (c.ord - 48) ** 2 }
    seen = {}
    while x != 1 && !seen[x]
        seen[x] = true
        x = x.digits.sum { |d| d ** 2 }
    end
    puts "#{s} #{x == 1 ? ':)' : ':('}"
end

File: src/01-easy/happy-numbers/solution.py
n = int(input())
for _ in range(n):
    s = input()
    seen = set()
    x = sum(int(c)**2 for c in s)
    while x != 1 and x not in seen:
        seen.add(x)
        x = sum(int(c)**2 for c in str(x))
    print(f"{s} :)" if x == 1 else f"{s} :(")
